pkg/yurthub/otaupdate/ota.go

```diff
@@ -0,0 +1,177 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package otaupdate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+const (
+	PodUpdatableAnnotation = "apps.openyurt.io/pod-updatable"
+)
+
+type PodStatus struct {
+	Namespace string
+	PodName   string
+	// TODO: whether need to display
+	// OldImgs    []string
+	// NewImgs    []string
+	Updatable bool
+}
+
+// GetPods return all daemonset pods' update information by PodStatus
+func GetPods(clientset kubernetes.Interface) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		podsStatus, err := getPodUpdateStatus(clientset)
+		if err != nil {
+			klog.Errorf("Get pods updatable status failed, %v", err)
+			returnErr(fmt.Errorf("Get daemonset's pods update status failed"), w, http.StatusInternalServerError)
+			return
+		}
+		klog.V(5).Infof("Got pods status list: %v", podsStatus)
+
+		// Successfully get daemonsets/pods update info
+		w.Header().Set("content-type", "text/json")
+		data, err := json.Marshal(podsStatus)
+		if err != nil {
+			klog.Errorf("Marshal pods status failed: %v", err.Error())
+			returnErr(fmt.Errorf("Get daemonset's pods update status failed: data transfer to json format failed."), w, http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		n, err := w.Write(data)
+		if err != nil || n != len(data) {
+			klog.Errorf("Write resp for request, expect %d bytes but write %d bytes with error, %v", len(data), n, err)
+		}
+	})
+}
+
+// UpdatePod update a specifc pod(namespace/podname) to the latest version
+func UpdatePod(clientset kubernetes.Interface) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+		namespace := params["ns"]
+		podName := params["podname"]
+
+		klog.Info("12", namespace, podName)
+		err, ok := applyUpdate(clientset, namespace, podName)
+		if err != nil {
+			returnErr(fmt.Errorf("Apply update failed"), w, http.StatusInternalServerError)
+			return
+		}
+		if !ok {
+			returnErr(fmt.Errorf("Pod is not-updatable"), w, http.StatusForbidden)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	})
+}
+
+// getDaemonsetPodUpdateStatus check pods annotation "apps.openyurt.io/pod-updatable"
+// to determine whether new version application is availabel
+func getPodUpdateStatus(clientset kubernetes.Interface) ([]*PodStatus, error) {
+	pods, err := clientset.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
```

Comment:
only pods on this node should be returned. not all of pods in the cluster.

Reply:
Should we do this by permission control? 
Maybe we need add a check in `/openyurt.io/v1/namespaces/{ns}/pods/{podname}/upgrade` as well?

Reply:
> Should we do this by permission control?

yes, we can only upgrade pods that run on the same node.